shell/app-shell/elements/sharing/single-user-context.js

```diff
@@ -0,0 +1,312 @@
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+*/
+
+import Xen from '../../../components/xen/xen.js';
+import Arcs from '../../../lib/arcs.js';
+import Firebase from '../../../lib/firebase.js';
+import {Field} from '../fb-data/Field.js';
+
+const log = Xen.logFactory('SingleUserContext', '#6f2453');
+
+const storage = new Arcs.StorageProviderFactory('shell');
+//
+const getSyntheticArcsStorageKey = arcid =>
+  `synthetic://arc/handles/${Firebase.storageKey}/arcs/${arcid}/serialization`;
+//
+const makeId = () => `id${Math.random()}`;
+const syntheticConnect = async storageKey =>
+  storage.connect(makeId(), null, getSyntheticArcsStorageKey(storageKey));
+const storeConnect = async (type, storageKey) =>
+  storage.connect(`id${Math.random()}`, type, storageKey);
+//
+const stores = {};
+const getSyntheticStore = async storageKey =>
+  stores[storageKey] || (stores[storageKey] = await syntheticConnect(storageKey));
+const getHandleStore = async ({type, storageKey}) =>
+  stores[storageKey] || (stores[storageKey] = await storeConnect(type, storageKey));
+//
+const snarfId = key => {
+  return key.split('/').pop();
+};
+
+const SingleUserContext = class {
+  constructor(context, userid, isProfile) {
+    this.isProfile = isProfile;
+    this.context = context;
+    this.userid = userid;
+    // we observe `arcid`s and `storageKey`s
+    this.observers = {};
+    // when we remove an arc from consideration, we have to unobserve storageKeys from that arc
+    // `handles` maps an arcid to an array of storageKeys to unobserve
+    this.handles = {};
+    this.field = new Field(null, `/users/${userid}`, userid, this._userSchema).activate();
+  }
+  dispose() {
+    // chuck all observers
+    Object.values(this.observers).forEach(({key}) => this._unobserve(key));
+    // chuck all data
+    this._removeUserEntities(this.context, this.userid, this.isProfile);
+  }
+  get _userSchema() {
+    return {
+      arcs: {
+        '*': {
+          $key: (parent, key, datum) => ({
+            $join: {
+              path: `/arcs/${parent}`,
+              schema: {
+                $changed: field => this._onShareChanged(field)
+              }
+            }
+          })
+        }
+      }
+    };
+  }
+  _onShareChanged(share) {
+    // TODO(sjmiles): schedule field processing to avoid re-entrancy
+    //log('SingleUserContext:onShareChanged', share);
+    const arcid = share.parent.key;
+    const userid = share.parent.path.split('/')[2];
+    const meta = share.data.metadata || Object;
+    if (share.disposed || meta.share < 2) {
```

Comment:
Didn't we have constants/enums for the `2` and `1`, maybe using those would be helpful